### PR TITLE
Rdm 1433

### PR DESCRIPTION
**Pull request by mario-paniccia**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RDM-1433

### Change description ###

- Updated Data Store to support async tasks
- Improved performance of /getUserProfile endpoint by parallelising external calls 
- Improved performance of /getUserProfile endpoint by retrieving only definition of jurisdicitions instead of all cases for all jurisdictions
- Simplified existing WireMockTests

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ X] No
```
⚠️ Asynchronous tasks for the moment are not compatible with request scoped beans, so for the moment can't be used together with cached repositories

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -176,6 +176,9 @@ dependencies {
     compile group: 'com.sun.mail', name: 'mailapi', version: '1.6.0'
     compile group: 'commons-lang', name: 'commons-lang', version: '2.6'
 
+    //simplifies handling of exceptions with Java CompletableFuture
+    compile group: 'org.zalando', name: 'faux-pas', version: '0.6.0'
```
Comment (mario-paniccia) [inline build.gradle]: handling of Exceptions in CompletableFuture involves lots of boilerplate. This library can help

File: src/main/java/uk/gov/hmcts/ccd/data/user/DefaultUserRepository.java [hunk 2]
```diff
@@ -58,50 +63,29 @@ public DefaultUserRepository(final ApplicationParams applicationParams,
     }
 
     @Override
-    public UserProfile getUserSettings() {
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/data/user/DefaultUserRepository.java]: getUserSettings has been renamed to getUserProfile.
Also, this is a method with complex orchestration logic more suitable for a Service than a Repository. 
I've introduced a UserService and move it there

File: src/main/java/uk/gov/hmcts/ccd/data/user/DefaultUserRepository.java [hunk 3]
```diff
@@ -127,8 +111,9 @@ public IDAMProperties getUserDetails() {
      * @param userId user id
      * @return UserDefault
      */
-    private UserDefault getUserDefaultSettings(final String userId) {
+    public UserDefault getUserDefaultSettings(final String userId) {
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/data/user/DefaultUserRepository.java]: this is used by UserService now

File: src/main/java/uk/gov/hmcts/ccd/data/user/UserRepository.java [hunk 4]
```diff
@@ -1,16 +1,20 @@
...
-
-import java.util.Set;
+import uk.gov.hmcts.ccd.domain.model.aggregated.UserDefault;
 
 public interface UserRepository {
-    UserProfile getUserSettings();
 
     IDAMProperties getUserDetails();
 
+    CompletableFuture<IDAMProperties> getUserDetailsAsync();
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/data/user/UserRepository.java]: methods exposed to UserService

File: src/main/java/uk/gov/hmcts/ccd/domain/service/aggregated/DefaultGetUserProfileOperation.java [hunk 5]
```diff
@@ -1,28 +1,24 @@
...
-    public DefaultGetUserProfileOperation(@Qualifier(CachedUserRepository.QUALIFIER) final UserRepository userRepository) {
-        this.userRepository = userRepository;
+    public DefaultGetUserProfileOperation(final UserService userService) {
+        this.userService = userService;
     }
 
     @Override
-    public UserProfile execute(final String userToken) {
-        final UserProfile userProfile = userRepository.getUserSettings();
-        if (userProfile == null) {
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/aggregated/DefaultGetUserProfileOperation.java]: userProfile will never be null

File: src/main/java/uk/gov/hmcts/ccd/domain/service/createcase/DefaultCreateCaseOperation.java [hunk 6]
```diff
@@ -44,7 +45,7 @@
     private final ValidateCaseFieldsOperation validateCaseFieldsOperation;
 
     @Inject
-    public DefaultCreateCaseOperation(@Qualifier(CachedUserRepository.QUALIFIER) final UserRepository userRepository,
+    public DefaultCreateCaseOperation(@Qualifier(DefaultUserRepository.QUALIFIER) final UserRepository userRepository,
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/createcase/DefaultCreateCaseOperation.java]: DefaultCreateCaseOperation does not retrieves userRoles, so does not need the cached repository

File: src/main/java/uk/gov/hmcts/ccd/endpoint/ui/UserProfileEndpoint.java [hunk 7]
```diff
@@ -30,7 +32,7 @@ public UserProfileEndpoint(@Qualifier(DefaultGetUserProfileOperation.QUALIFIER)
     @RequestMapping(value = "/caseworkers/{uid}/profile", method = RequestMethod.GET)
     @ApiOperation(value = "Get default setting for user")
     @ApiResponse(code = 200, message = "User default settings")
-    public UserProfile getUserProfile(@RequestHeader(value = AUTHORIZATION) final String authHeader) {
+    public Future<UserProfile> getUserProfile(@RequestHeader(value = AUTHORIZATION) final String authHeader) {
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/endpoint/ui/UserProfileEndpoint.java]: returning the Future makes the endpoint non-blocking

File: src/test/java/uk/gov/hmcts/ccd/WireMockBaseTest.java [hunk 8]
```diff
@@ -41,33 +41,9 @@ public void initMock() throws IOException {
...
 
-        stupApiCalls();
-
         ReflectionTestUtils.setField(applicationParams, "caseDefinitionHost", hostUrl);
         ReflectionTestUtils.setField(applicationParams, "uiDefinitionHost", hostUrl);
         ReflectionTestUtils.setField(applicationParams, "idamHost", hostUrl);
         ReflectionTestUtils.setField(applicationParams, "userProfileHost", hostUrl);
     }
-
-    protected void stupApiCalls() throws IOException {
```
Comment (mario-paniccia) [inline src/test/java/uk/gov/hmcts/ccd/WireMockBaseTest.java]: replaced with json stubbing as we have for all other stubs

File: src/main/java/uk/gov/hmcts/ccd/ApplicationParams.java [hunk 9]
```diff
@@ -55,6 +55,12 @@
     @Value("${definition.cache.ttl.secs}")
     private String definitionCacheTTLSecs;
 
+    @Value("${async.core.pool.size}")
+    private String asyncCorePoolSize;
+
+    @Value("${async.max.pool.size}")
+    private String asyncMaxPoolSize;
+
```
Comment (adr1ancho) [inline src/main/java/uk/gov/hmcts/ccd/ApplicationParams.java]: May I ask why Not have Integer here

File: src/main/java/uk/gov/hmcts/ccd/data/definition/DefaultCaseDefinitionRepository.java [hunk 10]
```diff
@@ -116,6 +121,31 @@ public UserRole getUserRoleClassifications(String userRole) {
...
+            final HttpEntity requestEntity = new HttpEntity(securityUtils.authorizationHeaders());
+            UriComponentsBuilder builder = UriComponentsBuilder.fromHttpUrl(applicationParams.jurisdictionDefURL());
+            return restTemplate.exchange(builder.build().encode().toUri(), HttpMethod.GET,
+                    requestEntity, new ParameterizedTypeReference<List<Jurisdiction>>() {}).getBody();
+        } catch (Exception e) {
+            LOG.warn("Error while retrieving jurisdictions definition", e);
+            if (e instanceof HttpClientErrorException
+                    && ((HttpClientErrorException)e).getRawStatusCode() == RESOURCE_NOT_FOUND) {
+                throw new ResourceNotFoundException("Resource not found when retrieving jurisdictions definition because of " + e.getMessage());
+            } else {
```
Comment (adr1ancho) [inline src/main/java/uk/gov/hmcts/ccd/data/definition/DefaultCaseDefinitionRepository.java]: Checkstyle should complain about line length

File: src/main/java/uk/gov/hmcts/ccd/data/user/DefaultUserRepository.java [hunk 11]
```diff
@@ -58,50 +63,29 @@ public DefaultUserRepository(final ApplicationParams applicationParams,
...
-        final WorkbasketDefault workbasketDefault = new WorkbasketDefault();
-        workbasketDefault.setJurisdictionId(userDefault.getWorkBasketDefaultJurisdiction());
-        workbasketDefault.setCaseTypeId(userDefault.getWorkBasketDefaultCaseType());
-        workbasketDefault.setStateId(userDefault.getWorkBasketDefaultState());
-        userProfile.getDefaultSettings().setWorkbasketDefault(workbasketDefault);
-        return userProfile;
+    public IDAMProperties getUserDetails() {
+        final HttpEntity requestEntity = new HttpEntity(securityUtils.userAuthorizationHeaders());
+        return restTemplate.exchange(applicationParams.idamUserProfileURL(), HttpMethod.GET, requestEntity, IDAMProperties.class).getBody();
     }
```
Comment (adr1ancho) [inline src/main/java/uk/gov/hmcts/ccd/data/user/DefaultUserRepository.java]: Checkstyle complains line length in new code

File: src/main/java/uk/gov/hmcts/ccd/data/user/DefaultUserRepository.java [hunk 12]
```diff
@@ -127,8 +111,9 @@ public IDAMProperties getUserDetails() {
      * @param userId user id
      * @return UserDefault
      */
-    private UserDefault getUserDefaultSettings(final String userId) {
+    public UserDefault getUserDefaultSettings(final String userId) {
         try {
+            LOG.debug("retrieving default user settings");
             final HttpEntity requestEntity = new HttpEntity(securityUtils.authorizationHeaders());
```
Comment (adr1ancho) [inline src/main/java/uk/gov/hmcts/ccd/data/user/DefaultUserRepository.java]: Just wondering whether you would like to debug also userId

LOG.debug("retrieving default user settings for userId {}", userId);

File: src/main/java/uk/gov/hmcts/ccd/data/user/UserService.java [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+                       JurisdictionMapper jurisdictionMapper) {
+        this.userRepository = userRepository;
+        this.caseDefinitionRepository = caseDefinitionRepository;
+        this.jurisdictionMapper = jurisdictionMapper;
+    }
+
+    public CompletableFuture<UserProfile> getUserProfileAsync() {
+
+        CompletableFuture<List<Jurisdiction>> jurisdictionDefsFuture = caseDefinitionRepository.getAllJurisdictionsAsync();
+        CompletableFuture<IDAMProperties> idamPropsFuture = userRepository.getUserDetailsAsync();
```
Comment (adr1ancho) [inline src/main/java/uk/gov/hmcts/ccd/data/user/UserService.java]: Checkstyle

File: src/main/java/uk/gov/hmcts/ccd/endpoint/exceptions/RestExceptionHandler.java [hunk 14]
```diff
@@ -42,11 +43,23 @@ public RestExceptionHandler(AppInsights appInsights) {
     @ExceptionHandler(Exception.class)
     @ResponseBody
     public ResponseEntity<HttpError> handleException(final HttpServletRequest request, final Exception exception) {
-        LOG.error(exception.getMessage(), exception);
+        LOG.error("data store request processing exception", exception);
         appInsights.trackException(exception);
```
Comment (adr1ancho) [inline src/main/java/uk/gov/hmcts/ccd/endpoint/exceptions/RestExceptionHandler.java]: What should be the level of LOG L32 said LOG.warn and here it is LOG.error

File: src/main/java/uk/gov/hmcts/ccd/ApplicationParams.java (hunk 9 above)
Comment (mario-paniccia) [reply src/main/java/uk/gov/hmcts/ccd/ApplicationParams.java]: was not aware Spring can handle automatically conversion from String to Integer. Will change this cheers

File: src/test/java/uk/gov/hmcts/ccd/endpoint/exceptions/RestExceptionHandlerIT.java [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+                .andExpect(status().isUnprocessableEntity())
+                .andExpect(jsonPath("$.error").value("Unprocessable Entity"));
+    }
+
+    @DisplayName("CompletionException not wrapping a ApiException should be unwrapped and handled as Exception")
+    @Test
+    public void testCompletionExceptionsCausedByNotApiException() throws Exception {
+
+        when(caseDetailsEndpoint.getDocumentsForEvent("jid", "ctid", "cid")).thenThrow(new CompletionException(new IOException("test exception")));
+
```
Comment (adr1ancho) [inline src/test/java/uk/gov/hmcts/ccd/endpoint/exceptions/RestExceptionHandlerIT.java]: Checkstyle

File: src/test/java/uk/gov/hmcts/ccd/endpoint/exceptions/RestExceptionHandlerIT.java [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+                .andExpect(status().isBadRequest())
+                .andExpect(jsonPath("$.error").value("Bad Request"));
+    }
+
+    @DisplayName("unknown exceptions should be handled as internal server error")
+    @Test
+    public void testUnknownExceptionsAreHandledAsInternalServerError() throws Exception {
+
+        when(caseDetailsEndpoint.getDocumentsForEvent("jid", "ctid", "cid")).thenThrow(new RuntimeException("test exception"));
+
```
Comment (adr1ancho) [inline src/test/java/uk/gov/hmcts/ccd/endpoint/exceptions/RestExceptionHandlerIT.java]: Checkstyle

File: src/main/java/uk/gov/hmcts/ccd/data/definition/DefaultCaseDefinitionRepository.java (hunk 10 above)
Comment (mario-paniccia) [reply src/main/java/uk/gov/hmcts/ccd/data/definition/DefaultCaseDefinitionRepository.java]: I know but we agreed line length of 120 chars is a bit too little so sometimes I ignore it
I really think with nowadays wide screens 120 chars is too small and the line wrapping that causes makes the code even less readable!

File: src/main/java/uk/gov/hmcts/ccd/data/user/DefaultUserRepository.java (hunk 12 above)
Comment (mario-paniccia) [reply src/main/java/uk/gov/hmcts/ccd/data/user/DefaultUserRepository.java]: ok sure

File: src/main/java/uk/gov/hmcts/ccd/data/definition/DefaultCaseDefinitionRepository.java (hunk 10 above)
Comment (adr1ancho) [reply src/main/java/uk/gov/hmcts/ccd/data/definition/DefaultCaseDefinitionRepository.java]: It is now open sourced and well

File: src/main/java/uk/gov/hmcts/ccd/endpoint/exceptions/RestExceptionHandler.java (hunk 14 above)
Comment (mario-paniccia) [reply src/main/java/uk/gov/hmcts/ccd/endpoint/exceptions/RestExceptionHandler.java]: we agreed to change this with Valentin. The thing is that handException handles unknown exceptions which for sure were not logged anywhere else in the code. 
handleApiException instead handles exceptions that probably have already been logged. So we changed to warn. 
But I've noticed lots of duplicated error logging, so we really should revisit our strategy

File: src/main/java/uk/gov/hmcts/ccd/CoreCaseDataApplication.java [hunk 17]
```diff
@@ -31,4 +43,26 @@ public static void main(String[] args) {
         SpringApplication.run(CoreCaseDataApplication.class, args);
     }
 
+    @Bean
+    public Executor asyncExecutor() {
+        //thread pool used for the execution of async tasks (methods marked with @Async)
+        ThreadPoolTaskExecutor executor = new ThreadPoolTaskExecutor();
+        executor.setCorePoolSize(applicationParams.getAsyncCorePoolSize());
+        executor.setMaxPoolSize(applicationParams.getAsyncMaxPoolSize());
+        executor.setQueueCapacity(500);
```
Comment (andrewfolga) [inline src/main/java/uk/gov/hmcts/ccd/CoreCaseDataApplication.java]: Should this not be a applicationParams config item?

Comment (mario-paniccia) [reply src/main/java/uk/gov/hmcts/ccd/CoreCaseDataApplication.java]: don't think it's necessary for everything to be part of applicationParams. I picked the settings that are likely to change more frequently. queue capacity is not one of them

**Pull request merged**